### PR TITLE
Validation for unsigned integer columns

### PR DESCRIPTION
**Pull request by dmromanov**
Unsigned Integer columns use "NonNegativeInteger" validation instead of a plain "Integer" validation.
For numeric and decimal columns "comparison" validator will be used. It will ensure that entered values are greater than zero.
Refs #439

### CODE REVIEW

**Comment by dereuromark** [comment]
ping @dmromanov

**Comment by dmromanov** [comment]
Hi! I am kind of stuck with making all tests pass. Tests heavily depend on DBMS used, so when tests pass on one DBMS they fail on others. Looking for a way to cover the code with tests while abstracting away from DBMS.
Advices or help are welcome.

File: src/Shell/Task/ModelTask.php [hunk 1]
```diff
@@ -707,11 +708,27 @@ public function fieldValidation($schema, $fieldName, array $metaData, $primaryKe
...
+            if ($metaData['unsigned']) {
+                $rules['nonNegativeInteger'] = [];
+            } else {
+                $rules['integer'] = [];
+            }
         } elseif ($metaData['type'] === 'float') {
             $rules['numeric'] = [];
+            if ($metaData['unsigned']) {
+                $rules['comparison'] = [
+                    '\Cake\Validation\Validation::COMPARE_GREATER_OR_EQUAL',
```
Comment (dereuromark) [inline src/Shell/Task/ModelTask.php]: Leading \ should not be in here.

File: tests/TestCase/Shell/Task/ModelTaskTest.php [hunk 2]
```diff
@@ -819,6 +830,24 @@ public function testGetValidation()
                 'scalar' => ['rule' => 'scalar', 'args' => []],
                 'allowEmpty' => ['rule' => 'allowEmpty', 'args' => []]
             ],
+            'rating' => [
+                'numeric' => ['rule' => 'numeric', 'args' => []],
+                'requirePresence' => ['rule' => 'requirePresence', 'args' => ["'create'"]],
+                'comparison' => ['rule' => 'comparison', 'args' => [
+                    '\Cake\Validation\Validation::COMPARE_GREATER_OR_EQUAL',
```
Comment (dereuromark) [inline tests/TestCase/Shell/Task/ModelTaskTest.php]: Same here

File: tests/TestCase/Shell/Task/ModelTaskAssociationDetectionTest.php [hunk 3]
```diff
@@ -166,6 +187,22 @@ public function testBakeAssociationDetectionCategoriesProductsTable()
      */
     public function testBakeAssociationDetectionOldProductsTable()
     {
+        $driver = ConnectionManager::get('test')->getDriver();
+        $this->skipIf($driver instanceof Mysql, 'Incompatible with sqlite');
```
Comment (dereuromark) [inline tests/TestCase/Shell/Task/ModelTaskAssociationDetectionTest.php]: Mysql?

File: tests/TestCase/Shell/Task/ModelTaskAssociationDetectionTest.php [hunk 4]
```diff
@@ -176,16 +213,50 @@ public function testBakeAssociationDetectionOldProductsTable()
      */
     public function testBakeAssociationDetectionProductVersionsTable()
     {
+        $driver = ConnectionManager::get('test')->getDriver();
+        $this->skipIf($driver instanceof Mysql, 'Incompatible with sqlite');
```
Comment (dereuromark) [inline tests/TestCase/Shell/Task/ModelTaskAssociationDetectionTest.php]: same

File: tests/TestCase/Shell/Task/ModelTaskAssociationDetectionTest.php (hunk 3 above)
Comment (dmromanov) [reply tests/TestCase/Shell/Task/ModelTaskAssociationDetectionTest.php]: Fixed.

File: tests/TestCase/Shell/Task/ModelTaskAssociationDetectionTest.php [hunk 5]
```diff
@@ -176,16 +213,50 @@ public function testBakeAssociationDetectionOldProductsTable()
...
+        $this->skipIf($driver instanceof Sqlite, 'Incompatible with sqlite');
+        $this->skipIf($driver instanceof Postgres, 'Incompatible with postgres');
+        $this->skipIf($driver instanceof Sqlserver, 'Incompatible with sqlserver');
         $this->_compareBakeTableResult('ProductVersions', __FUNCTION__);
     }
 
     /**
      * test checking if associations where built correctly for products.
      *
+     * @skip
```
[the remark this answers is not in the document]
Comment (dmromanov) [reply tests/TestCase/Shell/Task/ModelTaskAssociationDetectionTest.php]: I was experimenting with @skip annotation. Forgot to remove this :)

File: tests/TestCase/Shell/Task/ModelTaskTest.php [hunk 6]
```diff
@@ -949,6 +1111,75 @@ public function testGetValidationExcludeForeignKeys()
...
+                    0
+                ]],
+                'notEmpty' => ['rule' => 'notEmpty', 'args' => []],
+                'requirePresence' => ['rule' => 'requirePresence', 'args' => ["'create'"]],
+            ],
+            'score' => [
+                'decimal' => ['rule' => 'decimal', 'args' => []],
+                'comparison' => ['rule' => 'comparison', 'args' => [
+                    'Cake\Validation\Validation::COMPARE_GREATER_OR_EQUAL',
+                    0
```
[the remark this answers is not in the document]
Comment (dmromanov) [reply tests/TestCase/Shell/Task/ModelTaskTest.php]: Yes. It will fit more here.

[the remark this answers is not in the document]
Comment (dmromanov) [reply tests/TestCase/Shell/Task/ModelTaskTest.php]: Done.

**Pull request merged**